app/models/category.rb

```diff
@@ -0,0 +1,22 @@
+class Category < ApplicationRecord
+
+  belongs_to :parent_category, class_name: "Category", optional: true
+  has_many :products, dependent: :restrict_with_error
+  has_many :subcategories_products, through: :sub_categories, source: :products, dependent: :restrict_with_error
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id", dependent: :destroy
+
+  validates :name, presence: true
+  validates :name, uniqueness: true, if: :parent_category, allow_nil: true
+  validates :name, uniqueness: { scope: :parent_category_id }, allow_nil: true
+  validate NoChildOfSubCategoryValidator
+
+  scope :root, -> { where(parent_category_id: nil) }
+
+  private
```

Comment:
Should​ ​not​ ​be​ ​able​ ​to​ ​destroy​ ​category​ ​if​ ​any​ ​products​ ​are​ ​associated​ ​with​ ​its ​ ​​ ​​ ​​ ​sub_categories. where the callback for this?

Reply:
handled in association  dependent: :restrict_with_error